### DISCUSSION TITLE
how to change the elasticity disc heigh

### DISCUSSION THREAD

**Discussion by tina0425** [google-deepmind/mujoco#2178] [Asking for Help]
### Intro

Hi!

### My setup

Mujoco-3.2.4

### My question

I tried to use flexcomp (type-disc) to act like a IVD in the human spine, and I tried to increase the thickness of the disc to simulate different IVDs, But I tried to change the parameters in the file "mujoco-3.2.4-windows-x86_64\model\plugin\elasticity\pancake_flex.xml"  and found that there was no way to change the disc values in the z axis.

### Minimal model and/or code that explain my question

If you encountered the issue in a complex model, please simplify it as much as possible (while still reproducing the issue).

Model:
<details>
<summary>minimal XML</summary>

```XML

<flexcomp type="disc" count="9 9 1" spacing=".1 .1 .1" pos="0 0 .5"
              radius=".01" mass=".5" name="plate" dim="2">
      <contact condim="3" solref="0.01 1" solimp=".95 .99 .0001"/>
      <edge equality="true" damping="10"/>
      <plugin plugin="mujoco.elasticity.shell">
        <config key="poisson" value="0"/>
        <config key="thickness" value="8e-3"/>
        <!--Units are in Pa (SI)-->
        <config key="young" value="3e5"/>
      </plugin>
</flexcomp>

```

### Confirmations

- [x] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [ ] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
I think `<flexcomp radius>` does that?

**Reply by tina0425** [reply]
Thanks for your answer, I only want to increase the axial length, but modifying radius will change the shape of disc, such as from prototype to polygon, and radius must be smaller than spacing, which makes it impossible for me to modify the axial length only, because modifying spacing will result in a change in the disk size.

**Comment by quagla** [comment]
I assume you want axial deformations? In that case you need to use e volumetric (tetrahedral) mesh. If you only want radial deformations, you can use the `radius` as suggested by Yuval.

**Reply by tina0425** [reply]
Thanks for your reply, I will try to use volumetric mesh or use a mesh file. 
I really want your help, I am having trouble importing my mesh file both of msh and obj file, my msh model loads in and mujoco runs very slowly and keeps returning to the original position, I have tried to change many parameters, but nothing can solve my problem. When using obj files, this makes my disc model look like it's just a shell with no elastic stuff inside.
I use the code and grid file below, very eager for your reply

[soft_disc.zip](https://github.com/user-attachments/files/17541185/soft_disc.zip)

`<mujoco model="IVD">
  <option solver="CG" timestep="0.001" />
  <size memory="16G"/>


    <extension>
    <plugin plugin="mujoco.elasticity.shell" />
  </extension>


  <asset>
    <texture name="plane" type="2d" builtin="checker" rgb1=".2 .3 .4" rgb2=".1 0.15 0.2"
      width="512" height="512" mark="cross" markrgb=".8 .8 .8" />
    <material name="plane" reflectance="0.3" texture="plane" texrepeat="1 1" texuniform="true" />
    
  </asset>


  <worldbody>
    <geom name="floor" pos="0 0 0" size="0 0 .25" type="plane" material="plane" condim="3" />

    <flexcomp name="softbody1" type="gmsh" dim="3" rgba="0 .7 .7 1" radius=".0001" euler="0 0 0" pos="0 0 0" mass="10" file="asset/IVD-L1L2_vol.msh">
      
      <contact selfcollide="none" conaffinity="0" internal="false"/>
      <elasticity young="1e6" damping=".00" poisson="0.32"/>
    </flexcomp>

    <flexcomp name="softbody2" type="mesh" dim="2" rgba="0 .7 .5 1" radius=".0001" euler="0 0 0" pos="0 0 10" mass="10" file="asset/IVD-L1L2.obj">
      <contact selfcollide="none" internal="false"/>
      <edge equality="false" damping="1" />
      <plugin plugin="mujoco.elasticity.shell">
        <config key="poisson" value="0.5" />
        <config key="thickness" value="5e-6" />
        <config key="young" value="5e5" />       
      </plugin>
    </flexcomp>

    
    <light directional="true" diffuse=".2 .2 .2" specular="0 0 0" pos="0 0 5" dir="0 0 -1"
      castshadow="false" />
    <light directional="false" diffuse=".8 .8 .8" specular="0.3 0.3 0.3" pos="0 0 4" dir="0 0 -1" />
  </worldbody>
</mujoco>`

**Reply by quagla** [reply]
Yes, if you use an OBJ it will be just a shell with no axial deformation. It will be essentially rigid axially but it will have a thickness for collision purposes with a compliant contact model, so to some extent it could allow you to simulate some compliance in that way. But notice that with only a compliant contact you won't get any axial-radial coupling, so you will never get a bulging deformation from an axial loading.

It looks like your volumetric object is becoming unstable with that combination of Young modulus and time step. You have to reduce one or both of them in order to make the simulation stable, but unfortunately this will reduce the speed further. Unfortunately soft bodies are slow, there isn't much you can do besides trying to make a mesh with the fewest amount of vertices and even triangle size and possible (the time step is limited by the smallest triangle you have in the mesh).

**Reply by tina0425** [reply]
Well！ Thank you very much for your answer. Your answer has been very helpful to me.

 I tried to reduce the number of meshes to 600, which made my msh file run faster, but at the same time, the problem with the msh file is that as long as there is a parameter of young in the model, the simulation will be unstable, I am not sure whether it is still due to the number of grids. Perhaps you would agree to further reduce the possibility that the mesh number will have the opportunity to use the young parameter.
 
In addition, I also used an obj file with 600 meshes. The problem with this file was that it was also limited by young and thickness parameters. If young and thickness continued to increase, the model would be extremely unstable. I looked at #2073, and I found that young value in #2073  was larger than mine, whether it was also because my mesh number was too large.

`<worldbody>
    <geom name="floor" pos="0 0 0" size="0 0 .25" type="plane" material="plane" condim="3" />

    <flexcomp name="softbody1" type="gmsh" dim="3" rgba="0 .7 .7 1" radius=".0001" euler="0 0 0" pos="0 0 0" mass="10" file="asset/IVD-L1L2/IVDL1L2_vol.msh">
      
      <contact solref="0.01 1" solimp=".95 .99 .0001"  selfcollide="none" conaffinity="0" internal="false"/>
      <elasticity young="0" damping="1" poisson="0"/>
    </flexcomp>

    <flexcomp name="softbody2" type="mesh" dim="2" rgba="0 .7 .5 1" radius=".0001" euler="0 0 0" pos="50 0 0" mass="10" file="asset/IVD-L1L2/IVDL1L2.obj">
      <contact selfcollide="none" internal="false"/>
      <edge equality="true" solref="0.01 1" solimp=".95 .99 .0001" damping="1" />
      <plugin plugin="mujoco.elasticity.shell">
        <config key="poisson" value="0.5" />
        <config key="thickness" value="8e-3" />
        <config key="young" value="7e5" />       
      </plugin>
    </flexcomp>`
    
![fig1](https://github.com/user-attachments/assets/0d24fca0-b257-468c-bdb7-f979dddf6b34)


[soft_disc2.zip](https://github.com/user-attachments/files/17545118/soft_disc2.zip)
